From 0beccb1c27424aa6ceb5640be6ee79f90c651d1e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Horia=20Geant=C4=83?= <horia.geanta@nxp.com>
Date: Tue, 5 Dec 2017 11:14:10 +0200
Subject: [PATCH 343/767] crypto: caam/qi - fix FD congestion weight
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

When caam_qi_enqueue() is called, compound FD has already been swapped
to CAAM endianness, thus accesing it ("length" field in this case) by
CPU has to be done by first unswapping.

Fixes: ca982fae08a2 ("crypto: caam/qi - use QBMan (NXP) SDK driver")
Signed-off-by: Horia GeantÄƒ <horia.geanta@nxp.com>
[Xulin: Original patch taken from NXP LSDK-18.09.]
Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/crypto/caam/qi.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/crypto/caam/qi.c b/drivers/crypto/caam/qi.c
index 1d12f541cd11..1cf58b5cd779 100644
--- a/drivers/crypto/caam/qi.c
+++ b/drivers/crypto/caam/qi.c
@@ -110,7 +110,7 @@ int caam_qi_enqueue(struct device *qidev, struct caam_drv_req *req)
 
 	fd.cmd = 0;
 	fd.format = qm_fd_compound;
-	fd.cong_weight = req->fd_sgt[1].length;
+	fd.cong_weight = caam32_to_cpu(req->fd_sgt[1].length);
 	fd.addr = dma_map_single(qidev, req->fd_sgt, sizeof(req->fd_sgt),
 			      DMA_BIDIRECTIONAL);
 	if (dma_mapping_error(qidev, fd.addr)) {
@@ -578,15 +578,15 @@ static enum qman_cb_dqrr_result caam_rsp_fq_dqrr_cb(struct qman_portal *p,
 
 	fd = &dqrr->fd;
 	if (unlikely(fd->status)) {
-		u32 ssrc = status & JRSTA_SSRC_MASK;
-		u8 err_id = status & JRSTA_CCBERR_ERRID_MASK;
+               u32 ssrc = fd->status & JRSTA_SSRC_MASK; 
+               u8 err_id = fd->status & JRSTA_CCBERR_ERRID_MASK;
 
 		if (ssrc != JRSTA_SSRC_CCB_ERROR ||
 		    err_id != JRSTA_CCBERR_ERRID_ICVCHK)
 			dev_err(qidev, "Error: %#x in CAAM response FD\n", fd->status);
 	}
 
-	if (unlikely(fd->format != fd->format)) {
+       if (unlikely(fd->format != qm_fd_compound)) {
 		dev_err(qidev, "Non-compound FD from CAAM\n");
 		return qman_cb_dqrr_consume;
 	}
-- 
2.17.0

